File: apps/server/src/routes/ai.ts
import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth } from '../middleware/auth';
import { createValidationError, createInternalError, createNotFoundError } from '../middleware/error-handler';
import { logger } from '../utils/logger';
import { PlexService, resolvePlexToken } from '../services/plex';
import { finalizeImportResult } from '../services/import';
import { updateNotification } from '../services/job-notifications';
import { enqueueAction } from '../services/action-queue';
import { externalLimiter } from '../services/task-queues';
import axios from 'axios';

const router = Router();

// Gemini API configuration
const GEMINI_API_BASE = 'https://generativelanguage.googleapis.com/v1';
const GEMINI_MODEL = 'gemini-2.0-flash-exp';

// Grok API configuration
const GROK_API_BASE = 'https://api.x.ai/v1';
const GROK_MODEL = 'grok-beta';

/**
 * POST /api/import/ai
 * Generate a playlist using AI based on a text prompt
 */
router.post('/ai', requireAuth, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = req.session.userId!;
    const db = req.dbService!;
    const { prompt, trackCount = 50, geminiApiKey, grokApiKey, aiProvider } = req.body;

    if (!prompt || typeof prompt !== 'string') {
      return next(createValidationError('prompt is required'));
    }

    // Validate trackCount
    const validTrackCount = Math.min(100, Math.max(10, Number(trackCount) || 50));

    // Get user and server info
    const user = db.getUserById(userId);
    if (!user) {
      return next(createNotFoundError('User not found'));
    }

    // Get settings
    const settings = db.getUserSettings(userId);
    const selectedProvider = aiProvider || settings.ai_provider || 'gemini';
    
    // Get appropriate API key based on provider
    let apiKey: string;
    if (selectedProvider === 'grok') {
      apiKey = grokApiKey || settings.grok_api_key || '';
      if (!apiKey) {
        return next(createValidationError('Grok API key is required. Please add it in Settings or provide it in the request.'));
      }
    } else {
      apiKey = geminiApiKey || settings.gemini_api_key || '';
      if (!apiKey) {
        return next(createValidationError('Gemini API key is required. Please add it in Settings or provide it in the request.'));
      }
    }

    const userServer = db.getUserServer(userId);
    if (!userServer) {
      return next(createValidationError('No server selected. Please select a server first.'));
    }

    if (!userServer.library_id) {
      return next(createValidationError('No library selected. Please go to Settings and select a music library first.'));
    }

    // Initialize Plex service
    const plexService = new PlexService(userServer.server_url, resolvePlexToken(user, userServer));

    logger.info('AI playlist generation started', { userId, prompt, provider: selectedProvider });

    // Queued rather than run immediately: the rest of this (AI calls +
    // per-query Plex searches + playlist creation) can take a while, and
    // running an unbounded number of these across users at once is exactly
    // what the action queue caps. Track it in the notification bell instead
    // of holding the request open, same as every other import trigger (see
    // runNewImportAndFinalize's docs in services/import.ts).
    const { jobId, position } = enqueueAction(userId, 'AI Generated Playlist', async (notificationId) => {
     try {
      updateNotification(userId, notificationId, { detail: 'Asking AI for track ideas...', progress: 0 });
      // Use selected AI to extract search queries from the prompt - external
      // LLM call, so it shares externalLimiter with scraping/cross-import.
      const searchQueries = await externalLimiter.run(() => selectedProvider === 'grok'
        ? getSearchQueriesFromGrok(prompt, apiKey)
        : getSearchQueriesFromGemini(prompt, apiKey));
      logger.info('AI generated search queries', { userId, provider: selectedProvider, queries: searchQueries, queryCount: searchQueries.length });

      updateNotification(userId, notificationId, { detail: 'Searching your Plex library...', progress: 25 });

      // Search for tracks matching the queries
      const allTracks: any[] = [];
      for (const query of searchQueries.slice(0, 10)) { // Limit to 10 queries
        try {
          logger.info('Searching for query', { query, libraryId: userServer.library_id });
          const tracks = await plexService.searchTrack(query, userServer.library_id);
          logger.info('Search results', { query, trackCount: tracks.length });
          allTracks.push(...tracks.slice(0, 5)); // Take top 5 from each search
        } catch (error) {
          logger.warn('Failed to search for query', { query, error });
        }
      }

      logger.info('Total tracks found', { totalTracks: allTracks.length });

      // Remove duplicates
      const uniqueTracks = Array.from(
        new Map(allTracks.map(t => [t.ratingKey, t])).values()
      );

      // Shuffle and limit to requested track count
      const shuffled = uniqueTracks.sort(() => Math.random() - 0.5).slice(0, validTrackCount);

      // Generate playlist name from prompt using selected AI
      const playlistName = await externalLimiter.run(() => selectedProvider === 'grok'
        ? generatePlaylistNameWithGrok(prompt, apiKey)
        : generatePlaylistNameWithGemini(prompt, apiKey));
      updateNotification(userId, notificationId, { title: playlistName, detail: 'Creating Plex playlist...', progress: 75 });

      // Format as matched tracks
      const matched = shuffled.map(track => ({
        title: track.title,
        // originalTitle is the real per-track artist; grandparentTitle is just the
        // album/folder artist and can be wrong for compilations/soundtracks.
        artist: track.originalTitle || track.grandparentTitle || 'Unknown Artist',
        album: track.parentTitle || track.album,
        matched: true,
        plexRatingKey: track.ratingKey,
        score: 100,
      }));

      logger.info('AI playlist generated', { userId, trackCount: matched.length, playlistName });

      await finalizeImportResult(
        db,
        'ai',
        `AI: ${prompt}`,
        { id: userId, plex_token: user.plex_token },
        userServer,
        { matched, unmatched: [], playlistName, matchedCount: matched.length, totalCount: matched.length }
      );

      updateNotification(userId, notificationId, {
        title: playlistName,
        status: 'success',
        progress: 100,
        detail: `Added ${matched.length} tracks`,
      });
     } catch (error: any) {
      logger.error('Failed to generate AI playlist', { error: error.message, stack: error.stack, userId });
      const message = (error.response?.status === 401 || error.response?.status === 403)
        ? 'Invalid API key. Please check your API key and try again.'
        : (error.message || 'Failed to generate playlist');
      updateNotification(userId, notificationId, { status: 'error', detail: message });
     }
    }, 'import');

    res.json({ success: true, queued: true, jobId, position, message: 'Generating playlist...' });
  } catch (error: any) {
    logger.error('Failed to generate AI playlist', {
      error: error.message,
      stack: error.stack,
      userId: req.session.userId
    });

    // Check if it's an AI API error
    if (error.response?.status === 401 || error.response?.status === 403) {
      return next(createValidationError('Invalid API key. Please check your API key and try again.'));
    }

    next(createInternalError('Failed to generate playlist: ' + error.message));
  }
});

/**
 * POST /api/import/ai/test
 * Test the AI API key (Gemini or Grok)
 */
router.post('/ai/test', requireAuth, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = req.session.userId!;
    const db = req.dbService!;
    const { geminiApiKey, grokApiKey, aiProvider } = req.body;

    // Get settings
    const settings = db.getUserSettings(userId);
    const selectedProvider = aiProvider || settings.ai_provider || 'gemini';
    
    // Get appropriate API key
    let apiKey: string;
    if (selectedProvider === 'grok') {
      apiKey = grokApiKey || settings.grok_api_key || '';
      if (!apiKey) {
        return next(createValidationError('Grok API key is required'));
      }
    } else {
      apiKey = geminiApiKey || settings.gemini_api_key || '';
      if (!apiKey) {
        return next(createValidationError('Gemini API key is required'));
      }
    }

    logger.info('Testing AI API key', { userId, provider: selectedProvider });

    let testResponse;
    let response;
    
    if (selectedProvider === 'grok') {
      // Test Grok API
      testResponse = await axios.post(
        `${GROK_API_BASE}/chat/completions`,
        {
          model: GROK_MODEL,
          messages: [{
            role: 'user',
            content: 'Say "OK" if you can read this.'
          }],
          max_tokens: 10
        },
        {
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${apiKey}`
          },
          timeout: 10000
        }
      );
      response = testResponse.data.choices?.[0]?.message?.content || '';
    } else {
      // Test Gemini API
      testResponse = await axios.post(
        `${GEMINI_API_BASE}/models/${GEMINI_MODEL}:generateContent?key=${apiKey}`,
        {
          contents: [{
            parts: [{
              text: 'Say "OK" if you can read this.'
            }]
          }]
        },
        {
          headers: {
            'Content-Type': 'application/json'
          },
          timeout: 10000
        }
      );
      response = testResponse.data.candidates?.[0]?.content?.parts?.[0]?.text || '';
    }
    
    logger.info('AI API test successful', { userId, provider: selectedProvider, response });

    res.json({
      success: true,
      message: `${selectedProvider === 'grok' ? 'Grok' : 'Gemini'} API key is valid and working!`,
      testResponse: response
    });
  } catch (error: any) {
    logger.error('AI API test failed', { 
      error: error.message,
      status: error.response?.status,
      data: error.response?.data,
      userId: req.session.userId 
    });
    
    if (error.response?.status === 400 || error.response?.status === 401 || error.response?.status === 403) {
      return res.json({
        success: false,
        message: 'Invalid API key. Please check your API key.'
      });
    }
    
    if (error.code === 'ECONNABORTED' || error.code === 'ETIMEDOUT') {
      return res.json({
        success: false,
        message: 'Connection timeout. Please check your internet connection.'
      });
    }
    
    res.json({
      success: false,
      message: 'Failed to connect to AI API: ' + error.message
    });
  }
});

/**
 * Use Gemini AI to extract search queries from a user prompt
 */
async function getSearchQueriesFromGemini(prompt: string, apiKey: string): Promise<string[]> {
  try {
    const response = await axios.post(
      `${GEMINI_API_BASE}/models/${GEMINI_MODEL}:generateContent?key=${apiKey}`,
      {
        contents: [{
          parts: [{
            text: `You are a music expert. Given a user's description of a playlist they want to create, extract 5-10 specific search queries (artist names, song titles, genres, moods) that would help find matching tracks in a music library. Return ONLY a JSON array of strings, nothing else. Example: ["rock", "The Beatles", "upbeat", "80s pop", "dance music"]\n\nUser request: ${prompt}`
          }]
        }],
        generationConfig: {
          temperature: 0.7,
          maxOutputTokens: 200
        }
      },
      {
        headers: {
          'Content-Type': 'application/json'
        }
      }
    );

    const content = response.data.candidates?.[0]?.content?.parts?.[0]?.text || '[]';
    
    // Parse the JSON response
    try {
      const queries = JSON.parse(content);
      if (Array.isArray(queries) && queries.length > 0) {
        return queries.filter(q => typeof q === 'string' && q.length > 0);
      }
    } catch (parseError) {
      logger.warn('Failed to parse Gemini response as JSON, falling back to keyword extraction', { content });
    }

    // Fallback to simple keyword extraction
    return extractKeywords(prompt);
  } catch (error: any) {
    logger.error('Gemini API error', { 
      error: error.message,
      status: error.response?.status,
      data: error.response?.data 
    });
    
    // If Gemini fails, fall back to keyword extraction
    if (error.response?.status === 400 || error.response?.status === 403) {
      throw error; // Re-throw auth errors
    }
    
    return extractKeywords(prompt);
  }
}

/**
 * Use Gemini AI to generate a playlist name from a prompt
 */
async function generatePlaylistNameWithGemini(prompt: string, apiKey: string): Promise<string> {
  try {
    const response = await axios.post(
      `${GEMINI_API_BASE}/models/${GEMINI_MODEL}:generateContent?key=${apiKey}`,
      {
        contents: [{
          parts: [{
            text: `You are a creative playlist naming expert. Given a user's description of a playlist, create a short, catchy playlist name (2-6 words). Return ONLY the playlist name, nothing else. No quotes, no explanation.\n\nUser request: ${prompt}`
          }]
        }],
        generationConfig: {
          temperature: 0.8,
          maxOutputTokens: 50
        }
      },
      {
        headers: {
          'Content-Type': 'application/json'
        }
      }
    );

    const name = response.data.candidates?.[0]?.content?.parts?.[0]?.text?.trim() || '';
    
    if (name.length > 0 && name.length < 100) {
      // Remove quotes if present
      return name.replace(/^["']|["']$/g, '');
    }

    // Fallback
    return generatePlaylistName(prompt);
  } catch (error) {
    logger.warn('Failed to generate name with Gemini, using fallback', { error });
    return generatePlaylistName(prompt);
  }
}

/**
 * Extract keywords from a prompt (fallback method)
 */
function extractKeywords(prompt: string): string[] {
  // Remove common words
  const stopWords = new Set([
    'a', 'an', 'the', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for',
    'of', 'with', 'by', 'from', 'up', 'about', 'into', 'through', 'during',
    'create', 'make', 'playlist', 'songs', 'tracks', 'music', 'i', 'want',
    'like', 'that', 'this', 'some', 'any', 'all', 'would', 'could', 'should'
  ]);

  // Extract words
  const words = prompt
    .toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(word => word.length > 2 && !stopWords.has(word));

  // Return unique keywords
  return Array.from(new Set(words));
}

/**
 * Generate a playlist name from a prompt (fallback method)
 */
function generatePlaylistName(prompt: string): string {
  // Try to extract a meaningful name
  const cleaned = prompt
    .replace(/^(create|make|generate|build)\s+(a|an)?\s+(playlist\s+(of|with|for)?)?/i, '')
    .trim();

  if (cleaned.length > 0 && cleaned.length < 100) {
    // Capitalize first letter
    return cleaned.charAt(0).toUpperCase() + cleaned.slice(1);
  }

  // Fallback to generic name
  return 'AI Generated Playlist';
}

export default router;


/**
 * Use Grok AI to extract search queries from a user prompt
 */
async function getSearchQueriesFromGrok(prompt: string, apiKey: string): Promise<string[]> {
  try {
    const response = await axios.post(
      `${GROK_API_BASE}/chat/completions`,
      {
        model: GROK_MODEL,
        messages: [{
          role: 'user',
          content: `You are a music expert. Given a user's description of a playlist they want to create, extract 5-10 specific search queries (artist names, song titles, genres, moods) that would help find matching tracks in a music library. Return ONLY a JSON array of strings, nothing else. Example: ["rock", "The Beatles", "upbeat", "80s pop", "dance music"]\n\nUser request: ${prompt}`
        }],
        temperature: 0.7,
        max_tokens: 200
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`
        }
      }
    );

    const content = response.data.choices?.[0]?.message?.content || '[]';
    
    // Parse the JSON response
    try {
      const queries = JSON.parse(content);
      if (Array.isArray(queries) && queries.length > 0) {
        return queries.filter(q => typeof q === 'string' && q.length > 0);
      }
    } catch (parseError) {
      logger.warn('Failed to parse Grok response as JSON, falling back to keyword extraction', { content });
    }

    // Fallback to simple keyword extraction
    return extractKeywords(prompt);
  } catch (error: any) {
    logger.error('Grok API error', { 
      error: error.message,
      status: error.response?.status,
      data: error.response?.data 
    });
    
    // If Grok fails, fall back to keyword extraction
    if (error.response?.status === 400 || error.response?.status === 403) {
      throw error; // Re-throw auth errors
    }
    
    return extractKeywords(prompt);
  }
}

/**
 * Use Grok AI to generate a playlist name from a prompt
 */
async function generatePlaylistNameWithGrok(prompt: string, apiKey: string): Promise<string> {
  try {
    const response = await axios.post(
      `${GROK_API_BASE}/chat/completions`,
      {
        model: GROK_MODEL,
        messages: [{
          role: 'user',
          content: `You are a creative playlist naming expert. Given a user's description of a playlist, create a short, catchy playlist name (2-6 words). Return ONLY the playlist name, nothing else. No quotes, no explanation.\n\nUser request: ${prompt}`
        }],
        temperature: 0.8,
        max_tokens: 50
      },
      {
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`
        }
      }
    );

    const name = response.data.choices?.[0]?.message?.content?.trim() || '';
    
    if (name.length > 0 && name.length < 100) {
      // Remove quotes if present
      return name.replace(/^["']|["']$/g, '');
    }

    // Fallback
    return generatePlaylistName(prompt);
  } catch (error) {
    logger.warn('Failed to generate name with Grok, using fallback', { error });
    return generatePlaylistName(prompt);
  }
}

File: internal/services/ai/ai.go
// Package ai ports the plain-REST Gemini/Grok clients from routes/ai.ts
// (apps/server/src/routes/ai.ts:14-20,293-343,348-383,432-521) - no SDK for
// either provider, just net/http + encoding/json. Both providers do two
// things: turn a free-text playlist prompt into a handful of Plex search
// queries, and turn the same prompt into a short playlist name. Per-user API
// keys are stored in the users_settings table (gemini_api_key/grok_api_key),
// not env vars - verified there is no server-wide GEMINI_API_KEY/GROK_API_KEY
// in the current .env, so internal/config needs no changes for this.
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
	"unicode"
)

const (
	geminiModel = "gemini-2.0-flash-exp"
	grokModel   = "grok-beta"
)

// geminiAPIBase/grokAPIBase are vars, not consts, purely so tests can point
// them at a fake httptest server instead of the real APIs - same pattern
// and reasoning as auth.SetPlexAPIBaseForTest. Never reassigned outside
// tests.
var (
	geminiAPIBase = "https://generativelanguage.googleapis.com/v1"
	grokAPIBase   = "https://api.x.ai/v1"
)

// SetAPIBasesForTest points both providers at fake servers for the rest of
// the calling test, returning a restore func to defer.
func SetAPIBasesForTest(gemini, grok string) (restore func()) {
	prevGemini, prevGrok := geminiAPIBase, grokAPIBase
	geminiAPIBase, grokAPIBase = gemini, grok
	return func() { geminiAPIBase, grokAPIBase = prevGemini, prevGrok }
}

// AuthError mirrors the TS code's "re-throw auth errors instead of falling
// back to keyword extraction" behavior for 400/401/403 responses.
type AuthError struct {
	Status int
	Body   string
}

func (e *AuthError) Error() string {
	return fmt.Sprintf("ai provider auth error: status %d: %s", e.Status, e.Body)
}

var httpClient = &http.Client{Timeout: 10 * time.Second}

// ---- Gemini ----

type geminiRequest struct {
	Contents         []geminiContent `json:"contents"`
	GenerationConfig *geminiGenCfg   `json:"generationConfig,omitempty"`
}
type geminiContent struct {
	Parts []geminiPart `json:"parts"`
}
type geminiPart struct {
	Text string `json:"text"`
}
type geminiGenCfg struct {
	Temperature     float64 `json:"temperature"`
	MaxOutputTokens int     `json:"maxOutputTokens"`
}
type geminiResponse struct {
	Candidates []struct {
		Content struct {
			Parts []struct {
				Text string `json:"text"`
			} `json:"parts"`
		} `json:"content"`
	} `json:"candidates"`
}

func geminiGenerate(ctx context.Context, apiKey, prompt string, temperature float64, maxTokens int) (string, error) {
	body, _ := json.Marshal(geminiRequest{
		Contents:         []geminiContent{{Parts: []geminiPart{{Text: prompt}}}},
		GenerationConfig: &geminiGenCfg{Temperature: temperature, MaxOutputTokens: maxTokens},
	})
	url := fmt.Sprintf("%s/models/%s:generateContent?key=%s", geminiAPIBase, geminiModel, apiKey)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 400 || resp.StatusCode == 401 || resp.StatusCode == 403 {
		return "", &AuthError{Status: resp.StatusCode, Body: string(respBody)}
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("gemini api error: status %d: %s", resp.StatusCode, respBody)
	}

	var parsed geminiResponse
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return "", fmt.Errorf("gemini api: decode response: %w", err)
	}
	if len(parsed.Candidates) == 0 || len(parsed.Candidates[0].Content.Parts) == 0 {
		return "", nil
	}
	return parsed.Candidates[0].Content.Parts[0].Text, nil
}

// GetSearchQueriesFromGemini ports getSearchQueriesFromGemini (ai.ts:293-343).
// Auth errors are returned as *AuthError so callers can re-throw them instead
// of falling back, matching the TS behavior.
func GetSearchQueriesFromGemini(ctx context.Context, prompt, apiKey string) ([]string, error) {
	text, err := geminiGenerate(ctx, apiKey, searchQueryPrompt(prompt), 0.7, 200)
	if err != nil {
		var authErr *AuthError
		if isAuthError(err, &authErr) {
			return nil, err
		}
		return extractKeywords(prompt), nil
	}
	if queries := parseStringArray(text); len(queries) > 0 {
		return queries, nil
	}
	return extractKeywords(prompt), nil
}

// GeneratePlaylistNameWithGemini ports generatePlaylistNameWithGemini (ai.ts:348-383).
func GeneratePlaylistNameWithGemini(ctx context.Context, prompt, apiKey string) (string, error) {
	text, err := geminiGenerate(ctx, apiKey, namePrompt(prompt), 0.8, 50)
	if err != nil {
		// TS: logs a warning and falls back, never propagates.
		return generatePlaylistName(prompt), nil
	}
	if name := cleanName(text); name != "" {
		return name, nil
	}
	return generatePlaylistName(prompt), nil
}

// TestGemini ports the Gemini branch of POST /api/import/ai/test (ai.ts:233-252).
func TestGemini(ctx context.Context, apiKey string) (string, error) {
	return geminiGenerate(ctx, apiKey, `Say "OK" if you can read this.`, 0, 0)
}

// ---- Grok ----

type grokRequest struct {
	Model       string        `json:"model"`
	Messages    []grokMessage `json:"messages"`
	Temperature float64       `json:"temperature,omitempty"`
	MaxTokens   int           `json:"max_tokens,omitempty"`
}
type grokMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}
type grokResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
}

func grokChat(ctx context.Context, apiKey, content string, temperature float64, maxTokens int) (string, error) {
	body, _ := json.Marshal(grokRequest{
		Model:       grokModel,
		Messages:    []grokMessage{{Role: "user", Content: content}},
		Temperature: temperature,
		MaxTokens:   maxTokens,
	})
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, grokAPIBase+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 400 || resp.StatusCode == 401 || resp.StatusCode == 403 {
		return "", &AuthError{Status: resp.StatusCode, Body: string(respBody)}
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("grok api error: status %d: %s", resp.StatusCode, respBody)
	}

	var parsed grokResponse
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return "", fmt.Errorf("grok api: decode response: %w", err)
	}
	if len(parsed.Choices) == 0 {
		return "", nil
	}
	return parsed.Choices[0].Message.Content, nil
}

// GetSearchQueriesFromGrok ports getSearchQueriesFromGrok (ai.ts:432-481).
func GetSearchQueriesFromGrok(ctx context.Context, prompt, apiKey string) ([]string, error) {
	text, err := grokChat(ctx, apiKey, searchQueryPrompt(prompt), 0.7, 200)
	if err != nil {
		var authErr *AuthError
		if isAuthError(err, &authErr) {
			return nil, err
		}
		return extractKeywords(prompt), nil
	}
	if queries := parseStringArray(text); len(queries) > 0 {
		return queries, nil
	}
	return extractKeywords(prompt), nil
}

// GeneratePlaylistNameWithGrok ports generatePlaylistNameWithGrok (ai.ts:486-520).
func GeneratePlaylistNameWithGrok(ctx context.Context, prompt, apiKey string) (string, error) {
	text, err := grokChat(ctx, apiKey, namePrompt(prompt), 0.8, 50)
	if err != nil {
		return generatePlaylistName(prompt), nil
	}
	if name := cleanName(text); name != "" {
		return name, nil
	}
	return generatePlaylistName(prompt), nil
}

// TestGrok ports the Grok branch of POST /api/import/ai/test (ai.ts:212-232).
func TestGrok(ctx context.Context, apiKey string) (string, error) {
	return grokChat(ctx, apiKey, `Say "OK" if you can read this.`, 0, 10)
}

// ---- shared helpers ----

func isAuthError(err error, target **AuthError) bool {
	if ae, ok := err.(*AuthError); ok {
		*target = ae
		return true
	}
	return false
}

func searchQueryPrompt(prompt string) string {
	return "You are a music expert. Given a user's description of a playlist they want to create, " +
		"extract 5-10 specific search queries (artist names, song titles, genres, moods) that would " +
		`help find matching tracks in a music library. Return ONLY a JSON array of strings, nothing else. ` +
		`Example: ["rock", "The Beatles", "upbeat", "80s pop", "dance music"]` + "\n\nUser request: " + prompt
}

func namePrompt(prompt string) string {
	return "You are a creative playlist naming expert. Given a user's description of a playlist, " +
		"create a short, catchy playlist name (2-6 words). Return ONLY the playlist name, nothing else. " +
		"No quotes, no explanation.\n\nUser request: " + prompt
}

var quoteTrim = regexp.MustCompile(`^["']|["']$`)

func cleanName(text string) string {
	name := strings.TrimSpace(text)
	if len(name) == 0 || len(name) >= 100 {
		return ""
	}
	return quoteTrim.ReplaceAllString(name, "")
}

// parseStringArray mirrors the TS "JSON.parse then filter to non-empty
// strings" step; any parse failure or non-array result yields nil so the
// caller falls back to keyword extraction, same as the TS catch block.
func parseStringArray(text string) []string {
	var raw []interface{}
	if err := json.Unmarshal([]byte(text), &raw); err != nil {
		return nil
	}
	out := make([]string, 0, len(raw))
	for _, v := range raw {
		if s, ok := v.(string); ok && s != "" {
			out = append(out, s)
		}
	}
	return out
}

var stopWords = map[string]bool{
	"a": true, "an": true, "the": true, "and": true, "or": true, "but": true, "in": true, "on": true,
	"at": true, "to": true, "for": true, "of": true, "with": true, "by": true, "from": true, "up": true,
	"about": true, "into": true, "through": true, "during": true, "create": true, "make": true,
	"playlist": true, "songs": true, "tracks": true, "music": true, "i": true, "want": true, "like": true,
	"that": true, "this": true, "some": true, "any": true, "all": true, "would": true, "could": true,
	"should": true,
}

var nonWord = regexp.MustCompile(`[^\w\s]`)

// extractKeywords ports the fallback keyword extractor (ai.ts:388-406).
func extractKeywords(prompt string) []string {
	cleaned := nonWord.ReplaceAllString(strings.ToLower(prompt), " ")
	seen := map[string]bool{}
	var out []string
	for _, word := range strings.Fields(cleaned) {
		if len(word) <= 2 || stopWords[word] || seen[word] {
			continue
		}
		seen[word] = true
		out = append(out, word)
	}
	return out
}

var leadingPhrase = regexp.MustCompile(`(?i)^(create|make|generate|build)\s+(a|an)?\s*(playlist\s+(of|with|for)?)?`)

// generatePlaylistName ports the fallback name generator (ai.ts:411-424).
func generatePlaylistName(prompt string) string {
	cleaned := strings.TrimSpace(leadingPhrase.ReplaceAllString(prompt, ""))
	if len(cleaned) > 0 && len(cleaned) < 100 {
		r := []rune(cleaned)
		r[0] = unicode.ToUpper(r[0])
		return string(r)
	}
	return "AI Generated Playlist"
}
